### PR TITLE
[dev] Replace DP balance all-to-all rerouting with all-gather

### PR DESCRIPTION
**Pull request by xiaoyao0115**
## What changed

- Replace `DpBalancedScheduler` sample rerouting via `all_to_all_single` on the DP×CP group with per-field `all_gather_into_tensor` within each CP lane's DP group.
- Pad variable-size rank payloads to the largest DP-rank payload, gather in a fixed field order, and locally retain only samples assigned to the current DP×CP rank.
- Clone selected slices so returned samples do not retain the global gather buffer.
- Add mock, real NCCL, and DP×CP compatibility coverage, and update the scheduler documentation.

## Why

The first DP-balance all-to-all lazily initializes NCCL's fully connected P2P transport. At 64 ranks and 32 channels, its shareable receive/send allocations account for approximately:

```text
63 peers × 32 channels × (10 MiB + 2 MiB) = 23.625 GiB/GPU
```

Using the existing DP collective communicator for all-gather avoids that peer-per-channel transport allocation. In a 4-rank GB200 A/B smoke, the original path used 9804–10246 MiB/GPU and this path used 8652–9094 MiB/GPU, a 1152–1154 MiB/GPU reduction. This matches `3 peers × 32 channels × 12 MiB = 1.125 GiB/GPU`.

## Impact and tradeoff

This removes the large NCCL P2P transport memory overhead from DP-balance rerouting and remains compatible with CP siblings and dynamic CP rank assignment. The tradeoff is increased communication volume: every rank receives all DP shards before selecting its scheduled samples, so large-DP throughput should be profiled separately.

## Validation

- `python -m py_compile megatron/core/datasets/data_schedule.py megatron/core/datasets/data_schedule_utils.py tests/unit_tests/test_sequence_packing.py`
- `git diff --check`
- Four-rank GB200/NCCL run; all ranks passed:
  - `test_scheduler_reroute_uses_dp_all_gather`
  - `test_scheduler_reroute_dp_all_gather_distributed`
  - `test_scheduler_reroute_dp_all_gather_is_dcp_compatible` (DP=2, CP=2)
- Two-step GB200 THD/dynamic-CP smoke with DP=4 and dynamic CP sizes 1–4 completed training, validation, and test without skipped or NaN iterations.

## Additional observation

An EP=4 HybridEP+DCP A/B run left both the unmodified baseline and this branch in the same first-backward state for over ten minutes, so those jobs were stopped after collecting the memory comparison above. The matching baseline/feature behavior indicates this is separate from the rerouting change; the successful standard-dispatcher DCP smoke is used for the end-to-end compatibility result.

### CODE REVIEW

File: megatron/core/datasets/data_schedule_utils.py [hunk 1]
```diff
@@ -355,109 +355,103 @@ def reroute_samples_to_dcp_ranks(
...
+    for key in data_keys:
+        sample_numels = {
+            gid: 1 if key in scalar_keys else int(seq_len_by_gid[gid])
+            for gid in range(offset_values[-1])
+        }
+        rank_numels = [
+            sum(sample_numels[gid] for gid in range(offset_values[rank], offset_values[rank + 1]))
+            for rank in range(dp_size)
+        ]
+        max_rank_numel = max(rank_numels)
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply megatron/core/datasets/data_schedule_utils.py]: Thanks, fixed.

File: megatron/core/datasets/data_schedule_utils.py [hunk 2]
```diff
@@ -355,109 +355,103 @@ def reroute_samples_to_dcp_ranks(
...
+    before advancing to the next key so the full gather buffer can be freed.
     """
 
-    def _gid_to_src_rank(gid: int) -> int:
-        dp_src_rank = torch.bucketize(gid, offsets[1:] - 1)
-        dcp_rank = (
-            torch.distributed.get_process_group_ranks(dp_group)[dp_src_rank] // tp_group.size()
-        ) % dp_cp_group.size()
-        return dcp_rank
+    del tp_group, total_dcp_gpus  # Kept in the signature for existing callers.
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply megatron/core/datasets/data_schedule_utils.py]: Thanks, fixed.

File: megatron/core/datasets/data_schedule_utils.py [hunk 3]
```diff
@@ -355,109 +355,103 @@ def reroute_samples_to_dcp_ranks(
...
+        for key in (
+            "tokens",
+            "labels",
+            "loss_mask",
+            "position_ids",
+            "original_seq_len",
+            "padded_seq_len",
         )
+        if key in batch[0]
+    ]
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply megatron/core/datasets/data_schedule_utils.py]: Thanks, fixed.

File: megatron/core/datasets/data_schedule_utils.py [hunk 4]
```diff
@@ -355,109 +355,103 @@ def reroute_samples_to_dcp_ranks(
...
+            gather_input[: local_tensor.numel()].copy_(local_tensor)
+        else:
+            gather_input = local_tensor.contiguous()
 
-    recv_sample_with_id = {recv_id: recv_samples[i] for i, recv_id in enumerate(recv_ids_sorted)}
-    return recv_sample_with_id
+        if dp_size == 1:
+            gathered_tensor = gather_input
+        else:
+            gathered_tensor = local_tensor.new_empty(dp_size * max_rank_numel)
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply megatron/core/datasets/data_schedule_utils.py:464]: fixed

File: megatron/core/datasets/data_schedule_utils.py [hunk 5]
```diff
@@ -355,109 +355,103 @@ def reroute_samples_to_dcp_ranks(
...
+            torch.distributed.all_gather_into_tensor(gathered_tensor, gather_input, group=dp_group)
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply megatron/core/datasets/data_schedule_utils.py]: Thanks. We evaluated this and found the padding overhead is not significant for the intended workloads, so we prefer to keep the native all_gather_into_tensor path.

File: megatron/core/datasets/data_schedule_utils.py [hunk 6]
```diff
@@ -355,109 +355,103 @@ def reroute_samples_to_dcp_ranks(
...
+    assert (
+        local_ids == expected_local_ids
+    ), f"Local sample IDs {local_ids} do not match DP-rank range {expected_local_ids}."
+    assert len(batch) == len(
+        local_ids
+    ), f"Local batch size {len(batch)} does not match sample-ID count {len(local_ids)}."
+
+    recv_ids = sorted(
+        {gid for sample_id_group in sample_id_groups for gid in sample_id_group[dcp_rank]}
+    )
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply megatron/core/datasets/data_schedule_utils.py:415]: fixed

File: tests/unit_tests/test_sequence_packing.py [hunk 7]
```diff
@@ -79,6 +82,181 @@ def test_scheduler_sanitizes_thd_padding_values():
...
+        tp_group=torch.distributed.group.WORLD,
+        dp_cp_group=torch.distributed.group.WORLD,
+        total_dcp_gpus=world_size,
+    )
+
+    assert list(received) == [target_gid]
+    assert torch.equal(
+        received[target_gid]['tokens'],
+        torch.full((target_gid + 1,), target_gid, dtype=torch.int64, device=device),
+    )
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply tests/unit_tests/test_sequence_packing.py:248]: fixed

File: tests/unit_tests/test_sequence_packing.py [hunk 8]
```diff
@@ -79,6 +82,181 @@ def test_scheduler_sanitizes_thd_padding_values():
...
+            torch.tensor([1.0, 1.0, 0.0, 0.0]),
+            torch.tensor([0, 1, 2, 3]),
+            torch.tensor([4, 0], dtype=torch.int32),
+            torch.tensor([4, 0], dtype=torch.int32),
+        ]
+    )
+    gather_groups = []
+
+    def _all_gather_into_tensor(output, input_, group):
+        gather_groups.append(group)
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply tests/unit_tests/test_sequence_packing.py:130]: fixed

**Review by hxbai — APPROVED** [review]
LGTM

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/6378/commits/5d6ecb7f25ccc0e6ceba4815d8ca714dc9c95e82
